packages/ui/src/components/controls/NumberInput.tsx

```diff
@@ -0,0 +1,126 @@
+import clsx from 'clsx';
+import type {JSX} from 'preact';
+import {useCallback, useEffect, useRef, useState} from 'preact/hooks';
+import {useDocumentEvent} from '../../hooks';
+import {MouseButton, clamp} from '../../utils';
+import styles from './Controls.module.scss';
+
+type NumberInputProps = Omit<
+  JSX.HTMLAttributes<HTMLInputElement>,
+  'min' | 'max' | 'step' | 'value' | 'onChange' | 'onChangeCapture'
+> & {
+  min?: number;
+  max?: number;
+  step?: number;
+  value: number;
+  onChange?: (value: number) => void;
+  onChangeCapture?: (value: number) => void;
+};
+
+export function NumberInput({
+  min = -Infinity,
+  max = Infinity,
+  step = 1,
+  value = 0,
+  onChange,
+  onChangeCapture,
+  ...props
+}: NumberInputProps) {
+  const handleChange = onChangeCapture ? onChange : undefined;
+  const handleChangeCapture = onChangeCapture ?? onChange;
+
+  const inputRef = useRef<HTMLInputElement>();
+  const [currentValue, setCurrentValue] = useState(value);
+  const [active, setActive] = useState(false);
+  const [startX, setStartX] = useState(0);
+
+  useEffect(() => {
+    setCurrentValue(value);
+  }, [value]);
+
+  const handleDrag = useCallback(
+    (event: MouseEvent) => {
+      setCurrentValue(value => {
+        // ignore very large jumps in movement, as they often result from
+        // mouse acceleration issues
+        if (Math.abs(event.movementX) > 100) {
+          return value;
+        }
+
+        const newValue = clamp(min, max, value + event.movementX * step);
+        handleChange?.(newValue);
+        return newValue;
+      });
+    },
+    [min, max, step],
+  );
+
+  useDocumentEvent('pointerlockchange', () => {
+    if (document.pointerLockElement === inputRef.current) {
+      document.addEventListener('pointermove', handleDrag);
+    } else {
+      document.removeEventListener('pointermove', handleDrag);
+    }
+  });
+
+  return (
+    <input
+      type={'number'}
+      ref={inputRef}
+      min={min}
+      max={max}
+      step={step}
+      value={currentValue}
+      onChangeCapture={event => {
+        const value = parseInt((event.target as HTMLInputElement).value);
```

Comment:
In the component you use `inputRef.current` and `event.target` in somewhat randomly, I think we should stick to one.
```suggestion
        const value = parseInt(inputRef.current.value);
```